### ISSUE TITLE
Additional Dry Run Evaluation Requirements

### ISSUE THREAD

**Issue by np-soartech** [ITM-Soartech/itm-api#3]
# 1. Determine which alignment target is closest to (and furthest from) the delegator

Proposed change: Add an endpoint that takes a `session_id` (the delegators session) and returns the most and least aligned target ids

```
/api/v1/get_most_and_least_aligned?session_id=SESSION
```

Returns:

```josn
{
  "most_aligned": "TARGET_ID",
  "least_aligned": "TARGET_ID",
}
```

# 2. Computing alignment between a delegator and an ADM

One step in the evaluation requires computing the alignment between a delegator's decisions and an ADM's decisions on the evaluation scenarios.

There is currently an endpoint to compute the alignment between a target and a session (`/api/v1/alignment/session?session_id=SESSION&target_id=TARGET`). This endpoint can be used to compute the alignment between an ADM and a target and a delegator and a target but not between an ADM and a delegator (since neither are targets).

**Proposed Change**: Add an endpoint that computes the alignment between two sessions.

```
/api/v1/alignment/compare_sessions?session_id_1=SESSION1&session_id_2=SESSION2
```

Additionally, since alignment computations are no longer always used on targets, we should remove the `alignment_target_id` from the AlignmentResults object. 

Current ([source](https://github.com/ITM-Soartech/itm-api/blob/89df59b6bc2f161f072a29d8740da7d21ed754ed/itm-api.yaml#L456)):

```json
{
  "alignment_source": [
    {
      "scenario_id": "string",
      "probes": [
        "string"
      ],
      "additionalProp1": {}
    }
  ],
  "alignment_target_id": "string",
  "score": 1,
  "kdma_values": [
    {
      "kdma": "maximization",
      "value": 1,
      "additionalProp1": {}
    }
  ],
  "additionalProp1": {}
}
```

Proposed (`alignment_target_id` is removed)

```json
{
  "alignment_source": [
    {
      "scenario_id": "string",
      "probes": [
        "string"
      ],
      "additionalProp1": {}
    }
  ],
  "score": 1,
  "kdma_values": [
    {
      "kdma": "maximization",
      "value": 1,
      "additionalProp1": {}
    }
  ],
  "additionalProp1": {}
}
```

# 3. Support Distribution in Alignment Target

SoarTech is proposing to use a distribution as a target, this adds one additional object to send the distribution in the API

**Proposed change**: Add optional `kde` field to `AlignmentTarget` with type [sklearn.neighbors.KernelDensity](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KernelDensity.html). The object can be serialized and deserialized using `sklearn.externals.joblib`. Both `kdmas[].value` and `kdmas[].kde` are optional to support both teams.

Current API ([source](https://github.com/ITM-Soartech/itm-api/blob/89df59b6bc2f161f072a29d8740da7d21ed754ed/itm-api.yaml#L510))

```json
{
  "id": "string",
  "kdmas": [
    {
      "kdma": "maximization",
      "value": 1
    }
  ]
}
```

Proposed API:

```json
{
  "id": "string",
  "kdmas": [
    {
      "kdma": "maximization",
      "value": 1,
      "kde": "sklearn.neighbors.KernelDensity (serialized to base64 string for transport)"
    }
  ]
}
```

# 4. Minor Bookkeeping Change: Require `user_id` to be specified when creating a new session

The `new_session` endpoint already has an optional `user_id` field to specify a name for the session. It is currently optional. We propose we mark it as required to help with bookkeeping.

The `user_id` can be any string and could be an ADM id (e.g. `Kitware-Aligned-V2-Maximization`) or a delegator ID (or anything else)

**Comment by np-soartech** [comment]
# Update: 10 July 2024

KDMA profile computation and alignment score generation are currently packaged into one object. This tight coupling causes issues for instance with the new endpoint (change **2** above) which generates an alignment target between two sessions (i.e. an ADM and a delegator). Since the alignment results contains a single KDMA profile, it is not clear which of the two sessions the KDMA profile is for. Additionally, there is no way to compute the KDMA profile for a user without also asking for alignment between some target.

Proposal:

  - **Option 1**: (recommended) Remove computed KDMA profile (`kdma_values`) from `AlignmentResults & add a new standalone endpoint to compute a KDMA profile for a session
  - **Option 2**: Make `AlignmentResults.kdma_values` a list of KDMA Profiles which also contain session ids

Note: both options below will require adding an optional `kde` field to the KDMA profile and making `value` optional (see change 3 above). And both options will impact the new endpoint (change 2 above) as well as the existing alignment endpoints `/api/v1/alignment` since they also return an `AlignmentResults` object.


## Option 1 (recommended)


Changes

  - Remove computed KDMA profile (`kdma_values`) from `AlignmentResults` object
  - Add a new endpoint for retrieving the computed KDMA profile directly

Current proposed `AlignmentResults`:

```json
{
  "alignment_source": [
    {
      "scenario_id": "string",
      "probes": [
        "string"
      ],
    }
  ],
  "score": 1,
  "kdma_values": [
    {
      "kdma": "maximization",
      "value": 1,
    }
  ]
}
```

changes to

```json
{
  "alignment_source": [
    {
      "scenario_id": "string",
      "probes": [
        "string"
      ],
    }
  ],
  "score": 1
}
```

Add a new endpoint 

```
/api/v1/computed_kdma_profile?session=SESSION_ID
```

Which returns the object formerly in `AlignmentTarget.kdma_values`. Both `values` and `kde` are optional. One must be returned (see change 3 above)

```json
{
  "computed_kdma_profile": [
    {
      "kdma": "maximization",
      "value": 1,
      "kde": "sklearn.neighbors.KernelDensity (serialized to base64 string for transport)"
    }
  ]
}
```

## Option 2 

Changes

  - Change computed KDMA profile (`kdma_values`) from `AlignmentResults` to list of profiles with IDs
  - Add optional kde field
 
Current proposed `AlignmentResults`:

```json
{
  "alignment_source": [
    {
      "scenario_id": "string",
      "probes": [
        "string"
      ],
    }
  ],
  "score": 1,
  "computed_kdma_profiles": [
    {
      {
        "session_id": "SESSION_ID",
        "kdma": "maximization",
        "value": 1,
        "kde": "sklearn.neighbors.KernelDensity (serialized to base64 string for transport)"
      }
    }
  ]
}
```

**Comment by np-soartech** [comment]
# Update 11 July 2024

Currently we are providing one KDE per KDMA in a profile. Upon discussion with TA2, a request was made to include some metadata about the KDE along with the object. Specifically, a label of what the KDE means.

Additionally, we are proposing to change the KDE to a list of these objects so variants of the distribution can be used for training and evaluation analysis:

Current proposed KDMA profile:

```
  "computed_kdma_profiles": [
    {
      {
        "session_id": "SESSION_ID",
        "kdma": "maximization",
        "value": 1,
        "kde": "sklearn.neighbors.KernelDensity (serialized to base64 string for transport)"
      }
    }
```

Updated proposal

```
  "computed_kdma_profiles": [
    {
      {
        "session_id": "SESSION_ID",
        "kdma": "maximization",
        "value": 1,
        "kdes": {
            "<label>": {
              "kde": <sklearn.neighbors.KernelDensity (serialized to base64 string for transport)>
            }
        }
      }
    }
```

The KDE variants that can be made available at this time (detailed documentation for each variant coming soon):

  1.   `rawscores`: 1D KDE; built from raw scores of the probe choices the use made. This is the KDE that was used to compute alignment during the metric refinement eval.
  2.  `globalnorm`: 1d KDE; Same as `rawscores` but normalized across all available choices the user saw in the scenario (for instance, if the scenario only had a lowest value of 0.1 and a highest value of 0.9, this KDE would be stretched out to be from 0 to 1.
  3. `localnorm`: 1D KDE; Normalized for each probe individually. I.e. did the user pick the best possible option available?
  4.  `globalnormx_localnormy`: A 2D KDE constructed from `globalnorm` and `localnorm`. This is the KDE that will be used to compute alignment between two decision makers
